samples/net/cloud/aws_iot_mqtt/src/main.c

```diff
@@ -0,0 +1,494 @@
+		bo->retries_count++;
+	}
+#else
+	*next_backoff_ms = BACKOFF_CONST_MS;
+#endif
+}
+
+static int aws_client_try_connect(void)
+{
+	int ret;
+	uint32_t backoff_ms;
+	struct backoff_context bo;
+
+	backoff_context_init(&bo);
+
+	while (bo.retries_count <= bo.max_retries) {
+		ret = mqtt_connect(&client_ctx);
+		if (ret == 0) {
+			goto exit;
+		}
+
+		backoff_get_next(&bo, &backoff_ms);
+
+		LOG_ERR("Failed to connect: %d backoff delay: %u ms", ret, backoff_ms);
+		k_msleep(backoff_ms);
+	}
+
+exit:
+	return ret;
+}
+
+struct publish_payload {
+	uint32_t counter;
+};
+
+static const struct json_obj_descr json_descr[] = {
+	JSON_OBJ_DESCR_PRIM(struct publish_payload, counter, JSON_TOK_NUMBER),
+};
+
+static int publish(void)
+{
+	struct publish_payload pl = {.counter = messages_received_counter};
+
+	json_obj_encode_buf(json_descr, ARRAY_SIZE(json_descr), &pl, buffer, sizeof(buffer));
+
+	return publish_message(CONFIG_AWS_PUBLISH_TOPIC, strlen(CONFIG_AWS_PUBLISH_TOPIC), buffer,
+			       strlen(buffer));
+}
+
+void aws_client_loop(void)
+{
+	int rc;
+	int timeout;
+	struct pollfd fds;
+
+	aws_client_setup();
+
+	rc = aws_client_try_connect();
+	if (rc != 0) {
+		goto cleanup;
+	}
+
+	fds.fd = client_ctx.transport.tcp.sock;
+	fds.events = POLLIN | POLLHUP | POLLERR;
+
+	for (;;) {
+		timeout = mqtt_keepalive_time_left(&client_ctx);
+		rc = poll(&fds, 1u, timeout);
+		if (rc >= 0) {
+			if (fds.revents & POLLIN) {
+				rc = mqtt_input(&client_ctx);
+				if (rc != 0) {
+					LOG_ERR("Failed to read MQTT input: %d", rc);
+					break;
+				}
+			}
+
+			rc = mqtt_live(&client_ctx);
+			if ((rc != 0) && (rc != -EAGAIN)) {
+				LOG_ERR("Failed to live MQTT: %d", rc);
+				break;
+			}
+		} else {
+			LOG_ERR("poll failed: %d", rc);
+			break;
+		}
+
+		if (do_publish) {
+			do_publish = false;
+			publish();
+		}
+
+		if (do_subscribe) {
+			do_subscribe = false;
+			subscribe_topic();
+		}
+	}
+
+cleanup:
+	mqtt_disconnect(&client_ctx);
+
+	close(fds.fd);
+	fds.fd = -1;
+}
+
+int sntp_sync_time(void)
+{
+	int rc;
+	struct sntp_time now;
+	struct timespec tspec;
+
+	rc = sntp_simple(SNTP_SERVER, SYS_FOREVER_MS, &now);
+	if (rc == 0) {
+		tspec.tv_sec = now.seconds;
+		tspec.tv_nsec = ((uint64_t)now.fraction * (1000lu * 1000lu * 1000lu)) >> 32;
+
+		clock_settime(CLOCK_REALTIME, &tspec);
+
+		LOG_DBG("Acquired time from NTP server: %u", (uint32_t)tspec.tv_sec);
+
+	} else {
+		LOG_ERR("Failed to acquire SNTP, code %d\n", rc);
+	}
+	return rc;
+}
+
+static int resolve_broker_addr(struct sockaddr_in *broker)
+{
+	int ret;
+	struct zsock_addrinfo *ai = NULL;
+
+	const struct zsock_addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+		.ai_protocol = 0,
+	};
+
+	ret = zsock_getaddrinfo(CONFIG_AWS_ENDPOINT, AWS_BROKER_PORT, &hints, &ai);
+
+	if (ret == 0) {
+		memcpy(broker, ai->ai_addr, MIN(ai->ai_addrlen, sizeof(struct sockaddr_storage)));
+
+		char addr_str[INET_ADDRSTRLEN];
```

Comment:
It's more of a convention, but we aim to define variables at the top of the block.